fix(checkout): align address name field lengths with customer entity

SpiGAndromeda opened the pull request:
### 1. Why is this change necessary?
Customer registration fails when first names exceed 50 characters or last names exceed 60 characters, even though the `customer` table supports up to 255 characters. This data inconsistency between the `customer` and `customer_address` tables causes registration failures when customer data is copied to create address records during the registration process.

**Business Impact:** Customers with longer names cannot create accounts, leading to lost sales and poor user experience.

### 2. What does this change do, exactly?
This fix aligns the database schema and entity definitions across all related tables:

- **Database Migration**: Updates `customer_address` and `order_address` tables to support VARCHAR(255) for first_name and last_name fields (previously VARCHAR(50/60))
- **Entity Definitions**: Updates `CustomerAddressDefinition` and `OrderAddressDefinition` MAX_LENGTH constants from 50/60 to 255

### 3. Describe each step to reproduce the issue or behaviour.
- Navigate to register page
- Enter a first name that surpasses 50 characters or a last name that surpasses 60 characters
- Fill in the rest of the fields as you would normally do
- Click continue
- The shop will throw an error, telling you that the first (or last) name of the customer address entity being too long

**After this fix:**
Following the same steps results in successful registration with names up to 255 characters.

### 4. Please link to the relevant issues (if any).
Fixes #5882

### 5. Checklist
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation (N/A - schema change only)
- [x] This change has comments for package types, values, functions (where applicable)
- [x] I have read the contribution requirements

SpiGAndromeda committed:
fix(checkout): align address name field lengths with customer entity

SpiGAndromeda committed:
test(migration): add test for customer address field length migration

keulinho commented on tests/migration/Core/V6_8/Migration1756068709FixCustomerAddressNameFieldLengthsTest.php:
```diff
@@ -0,0 +1,61 @@
+        $migration = new Migration1756068709FixCustomerAddressNameFieldLengths();
+        static::assertSame(1756068709, $migration->getCreationTimestamp());
+
+        $migration->update($this->connection);
```
please execute the migration twice in the unit tests, that way we can make sure that it is safe to rerun the migration when something fails e.g mysql has gone away between the first and second query in your migraion

keulinho commented on src/Core/Migration/V6_8/Migration1756068709FixCustomerAddressNameFieldLengths.php:
```diff
@@ -0,0 +1,38 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Migration\V6_8;
```
Migrations in the 6_8 namespace will only be executed on 6.8 major line, so they won't be executed for now

Put the migration in the 6.7 namespace and it should be executed automatically

mitelg commented:
why was that on us and not the checkout team? :grimacing:

mitelg commented on src/Core/Migration/V6_8/Migration1756068709FixCustomerAddressNameFieldLengths.php:
```diff
@@ -0,0 +1,38 @@
+        ');
+    }
+
+    public function updateDestructive(Connection $connection): void
```
can be removed

mitelg commented on tests/migration/Core/V6_8/Migration1756068709FixCustomerAddressNameFieldLengthsTest.php:
```diff
@@ -0,0 +1,61 @@
+        static::assertSame('NO', $orderAddressColumns['last_name']['Null']);
+    }
+
+    public function testUpdateDestructiveIsEmpty(): void
```
can also be removed then. totally useless test :blush:

mitelg commented on changelog/_unreleased/2025-08-24-fix-customer-address-name-field-lengths.md:
```diff
@@ -0,0 +1,25 @@
+### Background
+Previously, customer registration would fail when first names exceeded 50 characters or last names exceeded 60 characters, even though the customer table supported up to 255 characters. This mismatch caused registration failures when customer data was copied to create address records.
+
+### Migration Required
```
this does not need to be stated. it will happen automatically while updating the shop

SpiGAndromeda committed:
fix(migration): move migration and test to V6_7

- Move migration from V6_8 to V6_7 directory
- Extend migration test with idempotency check

SpiGAndromeda committed:
test(checkout): fix test for registration with long names

SpiGAndromeda committed:
fix(migration): destructive update tests removed

SpiGAndromeda committed:
fix(changelog): "Migration Required" section removed.

SpiGAndromeda committed:
fix(bc-break): BC break added to .bc-exclude.php

mstegmeyer commented on tests/integration/Core/Checkout/Customer/SalesChannel/RegisterLongNamesTest.php:
```diff
@@ -0,0 +1,165 @@
+ * @internal
+ */
+#[Package('checkout')]
+class RegisterLongNamesTest extends TestCase
```
Please don't create a separate test case class. Just put the TestCases in the `RegisterRouteTest`.

mstegmeyer commented on src/Core/Migration/V6_7/Migration1756068709FixCustomerAddressNameFieldLengths.php:
```diff
@@ -0,0 +1,34 @@
+            ALTER TABLE `order_address`
+            MODIFY COLUMN `first_name` VARCHAR(255) COLLATE utf8mb4_unicode_ci NOT NULL,
+            MODIFY COLUMN `last_name` VARCHAR(255) COLLATE utf8mb4_unicode_ci NOT NULL
+        ');
```
Maybe we should split them up into 2 or 4 separate migrations to reduce processing length 🤔 (in case one fails & request limits

SpiGAndromeda replied:
I created four. With large address tables, that could be a problem doing all changes at once.

keulinho replied:
good hint regarding the performance impact here, as both tables (esp order adress when not cleaned up) can grow huge

Just curious about the performance implications overall for splitting the alter table into two seperate statements do we expect that it takes as long as doing both `modify column` in one statement or does it take twice as long? 

Because in that case overall migration processing time would be more important to me then the runtime of a single migration, because in the end the shop likely needs to be down for maintenance for that whole time

cc @jenskueper & @shyim do you know if it's faster to use a single `alter table` statement that mosifies two columns, rather then two seperate statements modifying one column each?

shyim replied:
if you split it I would split by table which are changed. If MySQL is rebuilding a table is better to have all alter table in one statement, so it's done just once. But I think changing varchar size is an instant change as it updates just the validation and does not a reallocation (as we dont use fixed char)

SpiGAndromeda committed:
test(checkout): move long name tests to registerroutetest (#12064)

SpiGAndromeda committed:
refactor(migration): split address field length migration (#12064)

SpiGAndromeda committed:
Merge branch 'trunk' into next-5882/fix-customer-address-name-field-length-mismatch

keulinho commented on tests/migration/Core/V6_7/Migration1756068709FixAddressNameFieldLengthsTest.php:
```diff
@@ -0,0 +1,85 @@
+#[CoversClass(Migration1756068710FixCustomerAddressLastNameLength::class)]
+#[CoversClass(Migration1756068711FixOrderAddressFirstNameLength::class)]
+#[CoversClass(Migration1756068712FixOrderAddressLastNameLength::class)]
+class Migration1756068709FixAddressNameFieldLengthsTest extends TestCase
```
when we split the original class we should split the test classes as well right?

SpiGAndromeda replied:
Isn't that getting a bit verbose? That would be 4 nearly identical classes. I mean there is no performance difference. But it you dont insist on it, I would leave it as it is.

SpiGAndromeda replied:
Splitted them.

shyim commented on tests/integration/Core/Checkout/Customer/SalesChannel/RegisterRouteTest.php:
```diff
@@ -1398,6 +1398,141 @@ public function testRegisterWithTooLongPassword(): void
         static::assertSame(':PASSWORD_IS_TOO_LONG', $error['detail']);
     }
 
+    public function testRegistrationWithLongNames(): void
```
do we really want to have integration tests for this? any integration test makes the ci little bit slower

SpiGAndromeda replied:
I gues I could reduce it to one test, testing 255 characters. But we didn't test it yet and it was a bug. So having at least one test seems reasonable.

mitelg replied:
what Soner probably wanted to know, if we could make it into a unit test instead of an expensive integration test

SpiGAndromeda replied:
Moved/rewrote them.

SpiGAndromeda committed:
test(checkout): move customer name length validation tests to unit level (#5882)

SpiGAndromeda committed:
chore: temp remove bc exclusions to avoid merge conflicts (#5882)

SpiGAndromeda committed:
Merge branch 'trunk' into next-5882/fix-customer-address-name-field-length-mismatch

SpiGAndromeda committed:
chore: revert temp removal of bc exclusions for name lengths (#5882)

SpiGAndromeda committed:
test: split address field length migration tests per class (#5882)

SpiGAndromeda committed:
Merge branch 'trunk' into next-5882/fix-customer-address-name-field-length-mismatch

SpiGAndromeda committed:
chore(checkout): temp remove name length tests for rebase (#5882)

SpiGAndromeda committed:
chore(checkout): revert all changes to RegisterRouteTest for clean rebase (#5882)

SpiGAndromeda committed:
Merge branch 'trunk' into next-5882/fix-customer-address-name-field-length-mismatch

SpiGAndromeda committed:
test(checkout): re-add customer name length validation tests (#5882)

SpiGAndromeda committed:
test(checkout): refactor register route tests for better maintainability (#5882)

keulinho commented on tests/migration/Core/V6_7/Migration1756068711FixOrderAddressFirstNameLengthTest.php:
```diff
@@ -0,0 +1,54 @@
+        static::assertSame('NO', $columns['first_name']['Null']);
+    }
+
+    public function testUpdateIsIdempotent(): void
```
probably you added this for the sake of "one test for one thing" but essentially this is exactly the same case as the other where you execute the migration only once, IMHO we should just keep one of the cases for the sake of pipeline performance

Additionally keep in mind that the migration tests run after shopware was already set up, to actually test the migration we need to manually change the column definition to be smaller and then execute the migration

SpiGAndromeda replied:
Lenght is now changed to the previous values before executing the migration. After that, its executed twice.

keulinho commented on tests/unit/Core/Checkout/Customer/SalesChannel/RegisterRouteTest.php:
```diff
@@ -91,21 +91,14 @@ public function testAccountType(): void
-            $this->createMock(DataValidator::class),
-            $this->createMock(DataValidationFactoryInterface::class),
-            $this->createMock(DataValidationFactoryInterface::class),
+        $register = $this->createRegisterRoute(
```
nice improvmenet to encapsulate this in a private method and reuse 👍 

we can make it even easier to read by using named params here, so we don't need to pass so many null values

SpiGAndromeda replied:
Refactored this.

SpiGAndromeda committed:
refactor(checkout): use named parameters in register route tests (#5882)

SpiGAndromeda committed:
test: improve migration tests for address field lengths (#5882)
